crates/wal-protocol/src/lib.rs

```diff
@@ -0,0 +1,136 @@
+}
+
+/// Header is set on every message
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct Header {
+    pub source: Source,
+    pub dest: Destination,
+    pub ack_mode: AckMode,
+}
+
+/// Identifies the source of a message
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum Source {
+    /// Message is sent from another partition processor
+    Processor {
+        /// if possible, this is used to reroute responses in case of splits/merges
+        partition_id: PartitionId,
+        partition_key: Option<PartitionKey>,
+        /// The current epoch of the partition leader. Readers should observe this to decide which
+        /// messages to accept. Readers should ignore messages coming from
+        /// epochs lower than the max observed for a given partition id.
+        leader_epoch: LeaderEpoch,
+        /// This represents the logical clock of the sender at the time of writing this command.
+        /// The receiver can (optionally) use this message to detect duplication.
+        sequence_number: Option<MessageIndex>, // -- or sender_sequencer_number
+        /// Which node is this message from?
+        node_id: PlainNodeId,
+    },
+    /// Message is sent from an ingress node
+    Ingress {
+        /// The identity of the sender node. Generational for fencing. Ingress is
+        /// stateless, so we shouldn't respond to requests from older generation
+        /// if a new generation is alive.
+        node_id: GenerationalNodeId,
+        /// Last config version observed by sender. If this is a newer generation
+        /// or an unknown ID, we might need to update our config.
+        nodes_config_version: ConfigVersion,
+        // If unset and AckMode is dedup, we use the node_id as a key.
+        dedup_key: Option<String>,
+        sequence_number: MessageIndex,
+    },
+    /// Message is sent from some control plane component (controller, cli, etc.)
+    ControlPlane {
+        // Reserved for future use.
+    },
+}
+
+/// Identifies the intended destination of the message
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum Destination {
+    /// Message is sent to partition processor
+    Processor { partition_key: PartitionKey },
+}
+
+/// Defines expectations from the sender on how to respond to this command.
+/// Some commands require an explicit acknowledge.
+#[derive(Debug, Clone, Copy)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum AckMode {
+    /// Acknowledge the command with a response. For partition processor-initiated messages, we
+    /// respond by writing on their WAL (bifrost log). For ingress, we respond acknowledge by
+    /// direectly sending a response to the node over the node-to-node stream.
+    Ack,
+    /// Received should deduplicate this message.
+    Dedup,
```

Comment:
I am wondering whether we need the `AckMode` anymore. So far `Acking` was used to tell senders that the PP has durably stored a message. With the log, this should be replaced by the log writer getting an ok from the loglets I assume. Deduplication could still be interesting if we want to notify a shuffle or a Kafka consumer to jump ahead in case they are replaying some events (but is kind of an optimization). One idea could be to simplify things by getting rid of the `AckMode`. We still want to keep the deduplication key and sequence number, though.

Reply:
That's a good point. The deduplication key and sequence number are in indeed in the header.